src/optimizer/group_expression.cpp

```diff
@@ -87,15 +87,21 @@ hash_t GroupExpression::Hash() const {
 bool GroupExpression::operator==(const GroupExpression &r) {
   bool eq = (op == r.Op());
 
-  for (size_t i = 0; i < child_groups.size(); ++i) {
-    eq = eq && (child_groups[i] == r.child_groups[i]);
+  auto left_groups = child_groups;
+  auto right_groups = r.child_groups;
+
+  std::sort(left_groups.begin(), left_groups.end());
+  std::sort(right_groups.begin(), right_groups.end());
+
+  for (size_t i = 0; i < left_groups.size(); ++i) {
+    eq = eq && (left_groups[i] == right_groups[i]);
   }
 
   return eq;
 }
 
 void GroupExpression::SetRuleExplored(Rule *rule) {
-  rule_mask_.set(rule->GetRuleIdx()) = true;
```

Comment:
In #1285 , you mentioned that  
>We created and added a handful of "plan_selection_tests" that test if the optimizer does the appropriate join ordering (it currently does not).

Is this line the reason that the optimizer is not doing it? Or there are other rules that cause this problem?  
For simple join, say 2 tables, is it possible that the "logical to physical" rule is applied first and the rest of the possible plan are pruned off, like early stopping?

Reply:
I am not sure about possible early pruning, however, without this change, rules such as Join Commutativity were never being applied